File: scripts/generate-kanji-audio.js
const fs = require('fs');
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '../.env.local') });

// All 79 N5 kanji from database
const kanjiList = [
  '一', '二', '九', '七', '人', '入', '八', '十', '三', '上',
  '下', '大', '女', '山', '川', '土', '千', '子', '小', '中',
  '五', '六', '円', '天', '日', '月', '木', '水', '火', '出',
  '右', '四', '左', '本', '白', '万', '今', '午', '友', '父',
  '北', '半', '外', '母', '休', '先', '名', '年', '気', '百',
  '男', '見', '車', '毎', '行', '西', '何', '来', '学', '金',
  '雨', '国', '東', '長', '前', '南', '後', '食', '校', '時',
  '高', '間', '話', '電', '聞', '語', '読', '生', '書'
].map(char => ({ character: char }));

const TTS_API_URL = 'https://texttospeech.googleapis.com/v1/text:synthesize';

// Create output directory
const outputDir = path.join(__dirname, '../public/audio/kanji');
if (!fs.existsSync(outputDir)) {
  fs.mkdirSync(outputDir, { recursive: true });
}

// Function to generate audio for a single kanji
async function generateAudio(kanji) {
  try {
    const { character } = kanji;
    // Use character code as filename
    const filename = `${character.charCodeAt(0)}.mp3`;
    const outputPath = path.join(outputDir, filename);

    // Skip if file already exists
    if (fs.existsSync(outputPath)) {
      console.log(`✓ Skipping ${character} (already exists)`);
      return;
    }

    const apiKey = process.env.GOOGLE_CLOUD_API_KEY;
    if (!apiKey) {
      throw new Error('GOOGLE_CLOUD_API_KEY not set');
    }

    // Construct the TTS request - use the kanji character for natural pronunciation
    const ttsRequest = {
      input: { text: character },
      voice: {
        languageCode: 'ja-JP',
        name: 'ja-JP-Wavenet-A',
        ssmlGender: 'FEMALE',
      },
      audioConfig: {
        audioEncoding: 'MP3',
        speakingRate: 0.85,
        pitch: 0.0,
      },
    };

    // Call Google TTS REST API
    const response = await fetch(`${TTS_API_URL}?key=${apiKey}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(ttsRequest),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(`API error ${response.status}: ${JSON.stringify(errorData)}`);
    }

    const data = await response.json();

    if (!data.audioContent) {
      throw new Error('No audio content in response');
    }

    // Write base64 audio to file
    const audioBuffer = Buffer.from(data.audioContent, 'base64');
    console.log(`Audio size for ${character}: ${audioBuffer.length} bytes`);
    
    fs.writeFileSync(outputPath, audioBuffer);
    console.log(`✓ Generated audio for ${character} -> ${filename}`);

    // Small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 200));

  } catch (error) {
    console.error(`✗ Failed to generate audio for ${kanji.character}:`, error.message);
  }
}

// Main function
async function main() {
  console.log('Starting kanji audio generation...');
  console.log(`Total kanji: ${kanjiList.length}`);
  console.log(`Output directory: ${outputDir}\n`);

  if (!process.env.GOOGLE_CLOUD_API_KEY) {
    console.error('Error: GOOGLE_CLOUD_API_KEY environment variable is not set');
    console.error('Make sure .env.local has GOOGLE_CLOUD_API_KEY=your-key');
    process.exit(1);
  }

  console.log('API Key found:', process.env.GOOGLE_CLOUD_API_KEY.substring(0, 10) + '...\n');

  // Generate audio for all kanji
  for (const kanji of kanjiList) {
    await generateAudio(kanji);
  }

  console.log('\n✓ Audio generation complete!');
  console.log(`Generated files in: ${outputDir}`);
}

main().catch(console.error);

File: scripts/generate_kanji_audio.py
#!/usr/bin/env python3
"""
Generate audio files for kanji primary_reading using Google TTS API (Chirp3 voice)
Saves MP3 files to public/audio/kanji/ using kanji character code as filename
"""

import json
import os
import base64
import time
import requests
from dotenv import load_dotenv

load_dotenv('.env.local')

GOOGLE_API_KEY = os.getenv('GOOGLE_CLOUD_API_KEY')
TTS_API_URL = 'https://texttospeech.googleapis.com/v1/text:synthesize'
OUTPUT_DIR = 'public/audio/kanji'

if not GOOGLE_API_KEY:
    print("Error: GOOGLE_CLOUD_API_KEY not found in .env.local")
    exit(1)


def generate_audio(text: str, output_path: str) -> bool:
    """Generate audio for text using Google TTS Chirp3 voice"""
    
    tts_request = {
        "input": {"text": text},
        "voice": {
            "languageCode": "ja-JP",
            "name": "ja-JP-Chirp3-HD-Leda",  # High quality Chirp3 voice (free tier)
        },
        "audioConfig": {
            "audioEncoding": "MP3",
            "speakingRate": 0.85,
            "pitch": 0.0,
        }
    }
    
    try:
        response = requests.post(
            f"{TTS_API_URL}?key={GOOGLE_API_KEY}",
            headers={"Content-Type": "application/json"},
            json=tts_request
        )
        
        if response.status_code != 200:
            print(f"  API Error: {response.status_code} - {response.text[:100]}")
            return False
        
        data = response.json()
        if "audioContent" not in data:
            print(f"  No audio content in response")
            return False
        
        # Decode base64 and save to file
        audio_bytes = base64.b64decode(data["audioContent"])
        with open(output_path, "wb") as f:
            f.write(audio_bytes)
        
        return True
        
    except Exception as e:
        print(f"  Error: {e}")
        return False


def main():
    # Load kanji data
    with open('public/data/kanji.json', 'r', encoding='utf-8') as f:
        kanji_data = json.load(f)
    
    print(f"Loaded {len(kanji_data)} kanji entries")
    
    # Filter kanji with primary_reading
    kanji_with_reading = [k for k in kanji_data if k.get('primary_reading')]
    print(f"Kanji with primary_reading: {len(kanji_with_reading)}")
    
    # Ensure output directory exists
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    
    # Track progress
    success_count = 0
    skip_count = 0
    fail_count = 0
    
    for i, kanji in enumerate(kanji_with_reading):
        char = kanji['character']
        reading = kanji['primary_reading']
        char_code = ord(char)
        output_path = f"{OUTPUT_DIR}/{char_code}.mp3"
        
        # Skip if file already exists
        if os.path.exists(output_path):
            skip_count += 1
            print(f"[{i+1}/{len(kanji_with_reading)}] {char} ({reading}) - SKIP (exists)")
            continue
        
        print(f"[{i+1}/{len(kanji_with_reading)}] {char} ({reading})", end=" ")
        
        if generate_audio(reading, output_path):
            success_count += 1
            print("✓")
        else:
            fail_count += 1
            print("✗")
        
        # Rate limiting - be respectful to the API
        time.sleep(0.1)
    
    print(f"\n=== Complete ===")
    print(f"Success: {success_count}")
    print(f"Skipped: {skip_count}")
    print(f"Failed: {fail_count}")


if __name__ == "__main__":
    main()
